File: SignalUI/UI/AttachmentSharing.m
//
// Copyright 2017 Signal Messenger, LLC
// SPDX-License-Identifier: AGPL-3.0-only
//

#import "AttachmentSharing.h"
#import <SignalCoreKit/Threading.h>
#import <SignalServiceKit/AppContext.h>
#import <SignalServiceKit/FunctionalUtil.h>
#import <SignalServiceKit/MIMETypeUtil.h>
#import <SignalServiceKit/TSAttachmentStream.h>
#import <YYImage/YYImage.h>

NS_ASSUME_NONNULL_BEGIN

@implementation AttachmentSharing

+ (void)showShareUIForAttachment:(TSAttachmentStream *)stream sender:(nullable id)sender
{
    OWSAssertDebug(stream);

    [self showShareUIForAttachments:@[ stream ] sender:sender];
}

+ (void)showShareUIForAttachment:(TSAttachmentStream *)stream
                          sender:(nullable id)sender
                      completion:(nullable AttachmentSharingCompletion)completion
{
    OWSAssertDebug(stream);

    [self showShareUIForAttachments:@[ stream ] sender:sender completion:completion];
}

+ (void)showShareUIForAttachments:(NSArray<TSAttachmentStream *> *)attachments sender:(nullable id)sender
{
    OWSAssertDebug(attachments.count > 0);

    [self showShareUIForActivityItems:attachments sender:sender completion:nil];
}

+ (void)showShareUIForAttachments:(NSArray<TSAttachmentStream *> *)attachments
                           sender:(nullable id)sender
                       completion:(nullable AttachmentSharingCompletion)completion
{
    OWSAssertDebug(attachments.count > 0);

    [self showShareUIForActivityItems:attachments sender:sender completion:completion];
}

+ (void)showShareUIForURL:(NSURL *)url sender:(nullable id)sender
{
    [self showShareUIForURL:url sender:sender completion:nil];
}

+ (void)showShareUIForURL:(NSURL *)url
                   sender:(nullable id)sender
               completion:(nullable AttachmentSharingCompletion)completion
{
    OWSAssertDebug(url);

    [AttachmentSharing showShareUIForActivityItems:@[
        url,
    ]
                                            sender:sender
                                        completion:completion];
}

+ (void)showShareUIForURLs:(NSArray<NSURL *> *)urls
                    sender:(nullable id)sender
                completion:(nullable AttachmentSharingCompletion)completion
{
    OWSAssertDebug(urls.count > 0);

    [AttachmentSharing showShareUIForActivityItems:urls sender:sender completion:completion];
}

+ (void)showShareUIForText:(NSString *)text sender:(nullable id)sender
{
    [self showShareUIForText:text sender:sender completion:nil];
}

+ (void)showShareUIForText:(NSString *)text
                    sender:(nullable id)sender
                completion:(nullable AttachmentSharingCompletion)completion
{
    OWSAssertDebug(text);

    [AttachmentSharing showShareUIForActivityItems:@[
        text,
    ]
                                            sender:sender
                                        completion:completion];
}

#ifdef DEBUG
+ (void)showShareUIForUIImage:(UIImage *)image
{
    OWSAssertDebug(image);

    [AttachmentSharing showShareUIForActivityItems:@[
        image,
    ]
                                            sender:nil
                                        completion:nil];
}
#endif

+ (void)showShareUIForActivityItems:(NSArray *)activityItems
                             sender:(nullable id)sender
                         completion:(nullable AttachmentSharingCompletion)completion
{
    OWSAssertDebug(activityItems);

    DispatchMainThreadSafe(^{
        UIActivityViewController *activityViewController =
            [[UIActivityViewController alloc] initWithActivityItems:activityItems applicationActivities:@[]];

        [activityViewController setCompletionWithItemsHandler:^(UIActivityType __nullable activityType,
            BOOL completed,
            NSArray *__nullable returnedItems,
            NSError *__nullable activityError) {
            if (activityError) {
                OWSLogInfo(@"Failed to share with activityError: %@", activityError);
            } else if (completed) {
                OWSLogInfo(@"Did share with activityType: %@", activityType);
            }

            if (completion) {
                DispatchMainThreadSafe(completion);
            }
        }];

        UIViewController *fromViewController = CurrentAppContext().frontmostViewController;
        while (fromViewController.presentedViewController) {
            fromViewController = fromViewController.presentedViewController;
        }

        if (activityViewController.popoverPresentationController) {
            if ([sender isKindOfClass:[UIBarButtonItem class]]) {
                activityViewController.popoverPresentationController.barButtonItem = (UIBarButtonItem *)sender;
            } else if ([sender isKindOfClass:[UIView class]]) {
                UIView *viewSender = (UIView *)sender;
                activityViewController.popoverPresentationController.sourceView = viewSender;
                activityViewController.popoverPresentationController.sourceRect = viewSender.bounds;
            } else {
                if (sender) {
                    OWSFailDebug(@"Unexpected sender of type %@", NSStringFromClass([sender class]));
                }

                // Centered at the bottom of the screen.
                CGRect sourceRect = CGRectZero;
                sourceRect.origin.x = fromViewController.view.center.x;
                sourceRect.origin.y = CGRectGetMaxY(fromViewController.view.frame);

                activityViewController.popoverPresentationController.sourceView = fromViewController.view;
                activityViewController.popoverPresentationController.sourceRect = sourceRect;
                activityViewController.popoverPresentationController.permittedArrowDirections = 0;
            }
        }

        OWSAssertDebug(fromViewController);
        [fromViewController presentViewController:activityViewController animated:YES completion:nil];
    });
}

@end

@interface TSAttachmentStream (AttachmentSharing) <UIActivityItemSource>

@end

@implementation TSAttachmentStream (AttachmentSharing)

// called to determine data type. only the class of the return type is consulted. it should match what
// -itemForActivityType: returns later
- (id)activityViewControllerPlaceholderItem:(UIActivityViewController *)activityViewController
{
    // HACK: If this is an image we want to provide the image object to
    // the share sheet rather than the file path. This ensures that when
    // the user saves multiple images to their camera roll the OS doesn't
    // asynchronously read the files and save them to them in a random
    // order. Note: when sharing a mixture of image and non-image data
    // (e.g. an album with photos and videos) the OS will still incorrectly
    // order the video items. I haven't found any way to work around this
    // since videos may only be shared as URLs.
    return self.isImage ? [UIImage new] : self.originalMediaURL;
}

// called to fetch data after an activity is selected. you can return nil.
- (nullable id)activityViewController:(UIActivityViewController *)activityViewController
                  itemForActivityType:(nullable UIActivityType)activityType
{
    if ([self.contentType isEqualToString:OWSMimeTypeImageWebp]) {
        return self.originalImage;
    }
    if (self.isAnimated) {
        return self.originalMediaURL;
    }
    return self.isImage ? self.originalImage : self.originalMediaURL;
}

@end

// YYImage does not specify that the sublcass still supports secure coding,
// this is required for anything that subclasses a class that supports secure
// coding. We do so here, otherwise copy / save will not work for YYImages

@interface YYImage (SecureCoding)

@end

@implementation YYImage (SecureCoding)

+ (BOOL)supportsSecureCoding
{
    return YES;
}

@end

NS_ASSUME_NONNULL_END

File: SignalUI/UI/AttachmentSharing.swift
//
// Copyright 2023 Signal Messenger, LLC
// SPDX-License-Identifier: AGPL-3.0-only
//

import SignalServiceKit
import YYImage

public class AttachmentSharing {

    private init() {}

    // MARK: -

    public static func showShareUI(
        for attachmentStream: TSAttachmentStream,
        sender: Any? = nil,
        completion: (() -> Void)? = nil
    ) {
        showShareUIForActivityItems([attachmentStream], sender: sender, completion: completion)
    }

    public static func showShareUI(
        for attachmentStreams: [TSAttachmentStream],
        sender: Any? = nil,
        completion: (() -> Void)? = nil
    ) {
        showShareUIForActivityItems(attachmentStreams, sender: sender, completion: completion)
    }

    // MARK: -

    public static func showShareUI(
        for url: URL,
        sender: Any? = nil,
        completion: (() -> Void)? = nil
    ) {
        showShareUIForActivityItems([url], sender: sender, completion: completion)
    }

    public static func showShareUI(
        for urls: [URL],
        sender: Any? = nil,
        completion: (() -> Void)? = nil
    ) {
        showShareUIForActivityItems(urls, sender: sender, completion: completion)
    }

    // MARK: -

    public static func showShareUI(
        for text: String,
        sender: Any? = nil,
        completion: (() -> Void)? = nil
    ) {
        showShareUIForActivityItems([text], sender: sender, completion: completion)
    }

    // MARK: -

    #if USE_DEBUG_UI

    public static func showShareUI(
        for image: UIImage,
        sender: Any? = nil,
        completion: (() -> Void)? = nil
    ) {
        showShareUIForActivityItems([image], sender: sender, completion: completion)
    }

    #endif

    // MARK: -

    private static func showShareUIForActivityItems(
        _ activityItems: [Any],
        sender: Any?,
        completion: (() -> Void)? = nil
    ) {
        DispatchMainThreadSafe {
            let activityViewController = UIActivityViewController(activityItems: activityItems, applicationActivities: nil)
            activityViewController.completionWithItemsHandler = { activityType, completed, returnedItems, activityError in
                if let activityError {
                    Logger.info("Failed to share with activityError: \(activityError)")
                } else if completed {
                    Logger.info("Did share with activityType: \(String(describing: activityType))")
                }

                if let completion {
                    DispatchMainThreadSafe(completion)
                }
            }

            var fromViewController = CurrentAppContext().frontmostViewController()
            while fromViewController?.presentedViewController != nil {
                fromViewController = fromViewController?.presentedViewController
            }

            if let popoverPresentationController = activityViewController.popoverPresentationController {
                if let barButtonItem = sender as? UIBarButtonItem {
                    popoverPresentationController.barButtonItem = barButtonItem
                } else if let uiView = sender as? UIView {
                    popoverPresentationController.sourceView = uiView
                    popoverPresentationController.sourceRect = uiView.bounds
                } else if let fromViewController {
                    if let sender {
                        owsFailDebug("Unexpected sender of type: \(sender.self)")
                    }

                    // Centered at the bottom of the screen.
                    let sourceRect = CGRect(
                        x: fromViewController.view.center.x,
                        y: fromViewController.view.frame.maxY,
                        width: 0,
                        height: 0
                    )

                    popoverPresentationController.sourceView = fromViewController.view
                    popoverPresentationController.sourceRect = sourceRect
                    popoverPresentationController.permittedArrowDirections = []
                }
            }

            fromViewController!.present(activityViewController, animated: true)
        }
    }
}

extension TSAttachmentStream: UIActivityItemSource {

    // called to determine data type. only the class of the return type is consulted. it should match what
    // -itemForActivityType: returns later
    public func activityViewControllerPlaceholderItem(_ activityViewController: UIActivityViewController) -> Any {
        // HACK: If this is an image we want to provide the image object to
        // the share sheet rather than the file path. This ensures that when
        // the user saves multiple images to their camera roll the OS doesn't
        // asynchronously read the files and save them to them in a random
        // order. Note: when sharing a mixture of image and non-image data
        // (e.g. an album with photos and videos) the OS will still incorrectly
        // order the video items. I haven't found any way to work around this
        // since videos may only be shared as URLs.
        if isImageMimeType {
            return UIImage()
        }
        return originalMediaURL as Any
    }

    // called to fetch data after an activity is selected. you can return nil.
    public func activityViewController(
        _ activityViewController: UIActivityViewController,
        itemForActivityType activityType: UIActivity.ActivityType?
    ) -> Any? {
        if contentType == OWSMimeTypeImageWebp {
            return originalImage
        }
        if isAnimatedMimeType == .animated {
            return originalMediaURL
        }
        if isImageMimeType {
            return originalImage
        }
        return originalMediaURL
    }
}

// YYImage does not specify that the sublcass still supports secure coding,
// this is required for anything that subclasses a class that supports secure
// coding. We do so here, otherwise copy / save will not work for YYImages
extension YYImage {
    open class override var supportsSecureCoding: Bool { true }
}
